indico/web/client/js/react/components/WTFRemoteSearchDropdown.jsx

```diff
@@ -0,0 +1,280 @@
+import React, {useCallback, useEffect, useMemo, useState} from 'react';
+import {Dropdown} from 'semantic-ui-react';
+
+import {Translate, PluralTranslate, Singular, Plural, Param} from 'indico/react/i18n';
+import {indicoAxios, handleAxiosError} from 'indico/utils/axios';
+import {makeAsyncDebounce} from 'indico/utils/debounce';
+
+const highlightSearch = (text, query = '') => {
+  const index = text.toLowerCase().indexOf(query);
+  if (!query || index === -1) {
+    return text;
+  }
+
+  return (
+    <>
+      {text.slice(0, index)}
+      <b>{text.slice(index, index + query.length)}</b>
+      {text.slice(index + query.length)}
+    </>
+  );
+};
+
+/**
+ * Dropdown which can dynamically query the backend
+ * for results using the search input. This is a replacement
+ * for the selectize.js library.
+ *
+ * It can be both controlled and uncontrolled.
+ * To make it controlled pass the value prop to it.
+ */
+export function RemoteSearchDropdown({
+  searchUrl,
+  searchMethod,
+  searchPayload,
+  minTriggerLength,
+  allowById,
+  preload,
+  valueField,
+  labelField,
+  searchField,
+  value: valueFromProps,
+  onChange: onChangeFromProps,
+  dropdownProps,
+}) {
+  const [searchQuery, setSearchQuery] = useState('');
+  const [options, setOptions] = useState([]);
+  const [loading, setLoading] = useState(false);
+
+  const isControlled = valueFromProps !== undefined;
+  const [internalValue, setInternalValue] = useState(null);
+  const value = isControlled ? valueFromProps : internalValue;
+
+  const debounce = useMemo(() => makeAsyncDebounce(200), []);
+
+  const getIdFromQuery = query => (query.match(/^#(\d+)$/) || {1: null})[1];
+
+  if (preload) {
+    // No need to limit the search length when all data is preloaded
+    minTriggerLength = 0;
+  }
+  const searchDisabled =
+    searchQuery.length < minTriggerLength && (!allowById || getIdFromQuery(searchQuery) === null);
+
+  const filterOptions = (opts, query) => {
+    if (searchDisabled) {
+      return [];
+    }
+
+    query = query.toLowerCase();
+    const id = getIdFromQuery(query);
+    return opts
+      .filter(opt => (id === null ? opt.search.includes(query) : +opt['friendly-id'] === +id))
+      .map(opt => ({
+        ...opt,
+        // this might lead to some unexpected results when
+        // searchField !== labelField, because the filtering is done
+        // on searchField but highlighting happens on labelField
+        text: highlightSearch(opt.text, query),
+      }));
+  };
+
+  const transformOptions = useCallback(
+    data => {
+      if (!Array.isArray(data)) {
+        return [];
+      }
+
+      const opts = data.map(item => ({
+        'key': item[valueField],
+        'value': item[valueField],
+        'text': item[labelField],
+        // extra values passed to <Dropdown.Item>
+        'search': item[searchField].toLowerCase(),
+        'friendly-id': item.friendly_id,
+      }));
+      return _.sortBy(opts, 'value');
```

Comment:
Is sorting by value correct? If this is for displaying the list of choices, sorting by text is likely the better choice (the value is probably an id, and thus not very meaningful for the user - it may not even be displayed anywhere)